client/src/components/TagField.js

```diff
@@ -11,38 +11,60 @@ import url from 'url';
 import debounce from 'debounce-promise';
 import PropTypes from 'prop-types';
 
-class TagField extends Component {
-  constructor(props) {
-    super(props);
+const TagField = ({
+  lazyLoad = false,
+  options,
+  creatable = false,
+  multi = false,
+  disabled = false,
+  labelKey = 'Title',
+  valueKey = 'Value',
+  SelectComponent = Select,
+  AsyncCreatableSelectComponent = AsyncCreatableSelect,
+  AsyncSelectComponent = AsyncSelect,
+  CreatableSelectComponent = CreatableSelect,
+  value: initialValue,
+  onChange,
+  optionUrl,
+  onBlur,
+  ...passThroughAttributesFromProps
+}) => {
+  const selectComponentRef = useRef(null);
+
+  const [initalState] = useState(initialValue || []);
+  const [hasChanges, setHasChanges] = useState(false);
+  const [value, setValue] = useState(initialValue);
+
+  useEffect(() => {
+    const element = selectComponentRef.current.inputRef;
+    const event = new Event('change', { bubbles: true });
+    element.dispatchEvent(event);
+  }, [hasChanges]);
 
-    this.selectComponentRef = React.createRef();
-
-    this.state = {
-      initalState: props.value ? props.value : [],
-      hasChanges: false,
-    };
-
-    if (!this.isControlled()) {
-      this.state = {
-        ...this.state,
-        value: props.value,
-      };
-    }
-
-    this.handleChange = this.handleChange.bind(this);
-    this.handleOnBlur = this.handleOnBlur.bind(this);
-    this.isValidNewOption = this.isValidNewOption.bind(this);
-    this.getOptions = this.getOptions.bind(this);
-    this.fetchOptions = debounce(this.fetchOptions, 500);
-  }
-
-  componentDidUpdate(previousProps, previousState) {
-    if (previousState.hasChanges !== this.state.hasChanges) {
-      const element = this.selectComponentRef.current.inputRef;
-      const event = new Event('change', { bubbles: true });
-      element.dispatchEvent(event);
-    }
-  }
+  /**
+   * Debounced function to fetch options from the server, optionally using the given string as a filter.
+   * The actual fetch is delayed by 500ms to avoid excessive requests while the user is typing.
+   *
+   * @param {string} input
+   * @return {Promise<Array<Object>>}
+   */
+  const fetchOptionsDebounced = useCallback(
```

Comment:
Had to move up as it's used by getOptions()

Reply:
We don't have a `fetchOptions` other than this one, so why rename it?
```suggestion
  const fetchOptions = useCallback(
```

Reply:
I had them split at one point. Updated.